src/algebraic_geometry/projective_spectrum/scheme.lean

```diff
@@ -120,36 +120,281 @@ def degree_zero_part {f : A} {m : ℕ} (f_deg : f ∈ 𝒜 m) : subring (away f)
+  erw [one_mul] at eq1,
+  simp only [mk_eq_mk', is_localization.eq] at eq1,
+  rcases eq1 with ⟨⟨_, ⟨M, rfl⟩⟩, eq1⟩,
+  erw [mul_one, mul_one] at eq1,
+  simp only [← subtype.val_eq_coe] at eq1,
+  refine set.ne_empty_iff_nonempty.mpr ⟨f^N * f^M, eq1.symm ▸ mul_mem_right _ _
+    (sum_mem _ (λ i hi, mul_mem_left _ _ _)), ⟨N+M, by rw pow_add⟩⟩,
+  generalize_proofs _ h,
+  exact (classical.some_spec h).1,
+end
+
+lemma no_intersection :
+  ((x.1.as_homogeneous_ideal.to_ideal : set A) ∩ (submonoid.powers f : set A)) = ∅ :=
+begin
+  by_contra rid,
+  rw [←ne.def, set.ne_empty_iff_nonempty] at rid,
+  choose g hg using rid,
+  obtain ⟨hg1, ⟨k, rfl⟩⟩ := hg,
+  by_cases k_ineq : 0 < k,
+  { erw x.1.is_prime.pow_mem_iff_mem _ k_ineq at hg1,
+    exact x.2 hg1 },
+  { erw [show k = 0, by linarith, pow_zero, ←ideal.eq_top_iff_one] at hg1,
+    apply x.1.is_prime.1,
+    exact hg1 },
+end
+
+/--The function between the basic open set `D(f)` in `Proj` to the corresponding basic open set in
+`Spec A⁰_f`. The fact that this function is continuous is proven in `Top_component.forward`.
+-/
+def to_fun : (Proj.T| (pbo f)) → (Spec.T (A⁰_ f_deg)) := λ x,
+⟨carrier f_deg x, carrier_ne_top f_deg x (no_intersection x), λ x1 x2 hx12, begin
+  haveI : decidable_eq (away f) := classical.dec_eq _,
+  rcases ⟨x1, x2⟩ with ⟨⟨x1, hx1⟩, ⟨x2, hx2⟩⟩,
+  induction x1 using localization.induction_on with data_x1,
+  induction x2 using localization.induction_on with data_x2,
+  rcases ⟨data_x1, data_x2⟩ with ⟨⟨a1, _, ⟨n1, rfl⟩⟩, ⟨a2, _, ⟨n2, rfl⟩⟩⟩,
+  rcases mem_carrier.clear_denominator f_deg x hx12 with ⟨c, N, acd, eq1⟩,
+  simp only [degree_zero_part.mul_val, localization.mk_mul, mul_one] at eq1,
+  simp only [mk_eq_mk', is_localization.eq] at eq1,
+  rcases eq1 with ⟨⟨_, ⟨M, rfl⟩⟩, eq1⟩,
+  erw [mul_one] at eq1,
+  simp only [← subtype.val_eq_coe, submonoid.coe_mul] at eq1,
+
+  rcases x.1.is_prime.mem_or_mem (show a1 * a2 * f ^ N * f ^ M ∈ _, from _) with h1|rid2,
+  rcases x.1.is_prime.mem_or_mem h1 with h1|rid1,
+  rcases x.1.is_prime.mem_or_mem h1 with h1|h2,
+  { left,
+    rw mem_carrier_iff,
+    simp only [show (mk a1 ⟨f ^ n1, _⟩ : away f) = mk a1 1 * mk 1 ⟨f^n1, ⟨n1, rfl⟩⟩,
+     by erw [localization.mk_mul, mul_one, one_mul]],
+    exact ideal.mul_mem_right _ _ (ideal.subset_span ⟨_, h1, rfl⟩), },
+  { right,
+    rw mem_carrier_iff,
+    simp only [show (mk a2 ⟨f ^ n2, _⟩ : away f) = mk a2 1 * mk 1 ⟨f^n2, ⟨n2, rfl⟩⟩,
+      by erw [localization.mk_mul, mul_one, one_mul]],
+    exact ideal.mul_mem_right _ _ (ideal.subset_span ⟨_, h2, rfl⟩), },
+  { exact false.elim (x.2 (x.1.is_prime.mem_of_pow_mem N rid1)), },
+  { exact false.elim (x.2 (x.1.is_prime.mem_of_pow_mem M rid2)), },
+  { rw eq1,
+    refine mul_mem_right _ _ (mul_mem_right _ _ (sum_mem _ (λ i hi, mul_mem_left _ _ _))),
+    generalize_proofs _ h,
+    exact (classical.some_spec h).1 },
+end⟩
+
+lemma preimage_eq (a : A) (n : ℕ)
```

Comment:
I think this statement is quite cryptic, and I think it could benefit from some comments describing it.

Reply:
added some documentation